src/proof/PGTransformationAlgorithms.cc

```diff
@@ -25,379 +25,324 @@ along with Periplo. If not, see <http://www.gnu.org/licenses/>.
-			if(n->getAnt2()!=NULL) n->getAnt2()->remRes( id );
-			if(isSetVisited1(id)) mpz_clear( safe_lit_set[id] );
-			removeNode(id);
-		}
-		else if(!n->isLeaf())
-		{
-			clauseid_t id1 = n->getAnt1()->getId();
-			clauseid_t id2 = n->getAnt2()->getId();
-			Var piv = n->getPivot();
-			int pos_piv = toInt(mkLit(piv,false));
-			int neg_piv = toInt(mkLit(piv,true));
-
-			short out = n->getAnt1()->hasOccurrenceBin(piv);
-			assert( out!=-1 );
-			// TODO guarantee positive occurrence pivot in ant1
-			bool ant1_has_pos_occ_piv = ( out == 0 );
-
-			// Enqueue antecedents
-			assert( n->getAnt1() ); assert( n->getAnt2() );
-			// 0 if no replacement, 1 if replacement by ant1, 2 if by ant2
-			short replace = 0;
-			// Check whether pivot present in all subgraph paths
-			if( mpz_tstbit( safe_lit_set[id], pos_piv ) )
-			{ if( ant1_has_pos_occ_piv ) replace = 1; else replace = 2; }
-			else if( mpz_tstbit( safe_lit_set[id], neg_piv) )
-			{ if( ant1_has_pos_occ_piv ) replace = 2; else replace = 1; }
-
-			// A node marked to be replaced is left with the replacing
-			// antecedent at ant1 and with ant2 set to NULL
-			if( replace == 1 )
-			{
-				n->getAnt2()->remRes( id );
-				n->setAnt2(NULL);
-			}
-			else if( replace == 2 )
-			{
-				n->getAnt1()->remRes( id );
-				n->setAnt1(n->getAnt2());
-				n->setAnt2(NULL);
-			}
+    recyclePivotsIter_RecyclePhase();
 
-			// Update antecedents pivots sets
-			// Propagate info to ant1
-			if( replace == 1 )
-			{
-				if(!isSetVisited1(id1))
-				{
-					mpz_init( safe_lit_set[id1] );
-					setVisited1(id1);
-					mpz_set(safe_lit_set[id1], safe_lit_set[id]);
-				}
-				else
-					mpz_and(safe_lit_set[id1], safe_lit_set[id1], safe_lit_set[id]);
-			}
-			// Propagate info to ant2
-			if( replace == 2 )
-			{
-				if(!isSetVisited1(id2))
-				{
-					mpz_init( safe_lit_set[id2] );
-					setVisited1(id2);
-					mpz_set(safe_lit_set[id2], safe_lit_set[id]);
-				}
-				else
-					mpz_and(safe_lit_set[id2], safe_lit_set[id2], safe_lit_set[id]);
-			}
-			// Propagate info to both antecedents if not replaced
-			if( replace == 0 )
-			{
-				// Set pivot bit for propagation
-				mpz_set( incr_safe_lit_set_1, safe_lit_set[id] );
-				mpz_set( incr_safe_lit_set_2, safe_lit_set[id] );
+    if (verbose() > 1) { std::cerr << "; Recycling end, restructuring begin" << std::endl; }
+    if (verbose() > 1) {
+        uint64_t mem_used = memUsed();
+        reportf("; Memory used after recycling, before restructuring: %.3f MB\n",
+            mem_used == 0 ? 0 : mem_used / 1048576.0);
+    }
 
-				if(ant1_has_pos_occ_piv)
-				{
-					mpz_setbit( incr_safe_lit_set_1, pos_piv);
-					mpz_setbit( incr_safe_lit_set_2, neg_piv);
-				}
-				else
-				{
-					mpz_setbit( incr_safe_lit_set_2, pos_piv);
-					mpz_setbit( incr_safe_lit_set_1, neg_piv);
-				}
+    auto hasBeenReplaced = [](ProofNode * node) { return node->getAnt1() != nullptr and node->getAnt2() == nullptr; };
```

Comment:
Not sure if this would be clearer...
```suggestion
    auto hasBeenReplaced = [](ProofNode * node) { return node->getAnt1() and not node->getAnt2(); };
```

Reply:
I was not sure myself, but let's go with your suggestion.

Reply:
Fixed.